### PR TITLE
Feature/multiple filters, fix dedup CLI config values & other fixes

### LINKED ISSUE #1164: `BinCommand` duplicates options of `AuditCommand` in `cargo-audit`
CLI flags in `cargo_audit::commands::audit::AuditCommand` are duplicated in `cargo_audit::commands::audit::binary_scanning::BinCommand`. This means that running help commands will produce:

<details>
<summary>$ cargo run -q -p cargo-audit -- audit -h</summary>

```
Audit Cargo.lock files for vulnerable crates

Usage: cargo audit [OPTIONS] [COMMAND]

Commands:
  bin   scan compiled binaries
  help  Print this message or the help of the given subcommand(s)

Options:
  -c, --color <COLOR>              color configuration: always, never (default: auto)
  -d, --db <DB>                    advisory database git repo path (default: ~/.cargo/advisory-db)
  -D, --deny <DENY>                exit with an error on: warnings (any), unmaintained, unsound, yanked
  -f, --file <FILE>                Cargo lockfile to inspect (or `-` for STDIN, default: Cargo.lock)
      --ignore <ADVISORY_ID>       Advisory id to ignore (can be specified multiple times)
      --ignore-source              Ignore sources of packages in Cargo.toml, matching advisories regardless of source
  -n, --no-fetch                   do not perform a git fetch on the advisory DB
      --stale                      allow stale database
      --target-arch <TARGET_ARCH>  filter vulnerabilities by CPU (default: no filter)
      --target-os <TARGET_OS>      filter vulnerabilities by OS (default: no filter)
  -u, --url <URL>                  URL for advisory database git repo
  -q, --quiet                      Avoid printing unnecessary information
      --json                       Output report in JSON format
  -h, --help                       Print help
  -V, --version                    Print version
```
</details>

and

<details>
<summary>$ cargo run -q -p cargo-audit -- audit bin -h</summary>

```
scan compiled binaries

Usage: cargo audit bin [OPTIONS] <BINARY_PATHS>...

Arguments:
  <BINARY_PATHS>...  Paths to the binaries to be scanned

Options:
  -c, --color <COLOR>              color configuration: always, never (default: auto)
  -d, --db <DB>                    advisory database git repo path (default: ~/.cargo/advisory-db)
  -D, --deny <DENY>                exit with an error on: warnings (any), unmaintained, unsound, yanked
      --ignore <ADVISORY_ID>       Advisory id to ignore (can be specified multiple times)
      --ignore-source              Ignore sources of packages in the audit data, matching advisories regardless of source
  -n, --no-fetch                   do not perform a git fetch on the advisory DB
      --stale                      allow stale database
      --target-arch <TARGET_ARCH>  filter vulnerabilities by CPU (default: no filter)
      --target-os <TARGET_OS>      filter vulnerabilities by OS (default: no filter)
  -u, --url <URL>                  URL for advisory database git repo
  -q, --quiet                      Avoid printing unnecessary information
      --json                       Output report in JSON format
  -h, --help                       Print help (see more with '--help')
```
</details>

This in turn allows to run `cargo audit bin` with duplicated and possibly conflicting config values. For example you can run:

```
cargo run -q -p cargo-audit -- audit --url https://foo bin --url https://bar /path/to/binary
```

I think this is some mistake and should be fixed by removing duplicated options from `BinCommand`. This will be of course a breaking change, but this seams like a right thing to do. And I doubt this will break many people, due to how odd it would be for someone to rely on this.

I am currently implementing #1160 and can to this together, if my proposition is accepted.

### PR DESCRIPTION
**Pull request by aleksanderkrauze**
This PR:

- implements #1160
- fixes #1164
- fixes parsing of CLI `--color` flag

I have formulated a couple of questions whilst I was reading code and would like to ask them here.

Should `cargo-audit --color` support value `termcolor::ColorChoice::AlwaysAnsi`? Before I refactored `--color` flag, this value was not supported, so I left it that way. But should this remain?

After I removed duplicated flags from `BinCommand` I have noticed that `FixCommand` also duplicates flag `--file` of `AuditCommand`, which allows to specify path to `Cargo.lock` file. I wanted to remove it as well, and use `--file` passed to `AuditCommand`, when running `FixCommand`, but I don't think this is easy to do, because `FixCommand` implements `Runnable` and its method `run` does not allow to pass arguments from external context.

I have never before worked with `Abscissa` before, so I don't know what is the best possible solution. Maybe `FixCommand` doesn't need to implement `Command`? Maybe `--file` should be also added to `AuditConfig`? Maybe  `FixCommand.file` should be changed to `RefCell<Option<PathBuf>>`, skipped in clap, and then set in `AuditCommand::run` before it invokes `FixCommand::run`? I would be happy to hear a suggestion on how to implement it best. And I think it should be fixed, the same as duplicated arguments in `BinCommand` should have been fixed.

I would also like to do something with another issue I have created - #1161. As I understand, crate `platforms` is auto-generated, and so are values for target `Arch` and `OS`. But where exactly are they generated from? And how to point users to an exhaustive list of those values?

### CODE REVIEW

**Comment by Shnatsel** [comment]
Thanks! So CI is going to fail but it's not your fault - it also fails on master due to reliance on a yanked crate, #1156

I should probably fix that first so that we get working CI again.

**Comment by aleksanderkrauze** [comment]
Ok. I will wait for an update info and then I will rebase.

And I will be wanting concurrently for any review comments, or answers to the questions I asked.

**Comment by Shnatsel** [comment]
I've fixed CI in #1167, please rebase or merge

**Comment by Shnatsel** [comment]
> I have never before worked with `Abscissa` before, so I don't know what is the best possible solution. Maybe `FixCommand` doesn't need to implement `Command`? Maybe `--file` should be also added to `AuditConfig`? Maybe `FixCommand.file` should be changed to `RefCell<Option<PathBuf>>`, skipped in clap, and then set in `AuditCommand::run` before it invokes `FixCommand::run`? I would be happy to hear a suggestion on how to implement it best. And I think it should be fixed, the same as duplicated arguments in `BinCommand` should have been fixed.

@tarcieri any thoughts on the interaction between Clap and Abscissa here?

**Comment by tarcieri** [comment]
clap was retrofit into abscissa which previously used gumdrop as its command line arguments parser.

I don't know exactly what changes are needed. If someone wants to open a PR I can take a look.

**Comment by aleksanderkrauze** [comment]
Oh, sorry. I've missed your message, that I can now rebase. I apologize for the delay.

**Comment by aleksanderkrauze** [comment]
I have rebased.

What about my question regarding `FixCommand`? I would like to propose some solution, so it can be discussed as @tarcieri suggested, but I would prefer to do it after current changes are accepted. Or should this be split off to another PR?

**Comment by aleksanderkrauze** [comment]
@Shnatsel Hi. I don't want to sound pushy, but could someone perhaps review this PR, please? I have ideas how to dedup `--file` flag, but I would like to be firstly sure, that current patches are accepted.

**Comment by Shnatsel** [comment]
Oh! Thanks for the ping, the notification about the PR update got lost in my inbox. Let me take a look

**Comment by Shnatsel** [comment]
CI fails due to latest rustc adding more lints. I've fixed it in #1181; CI should pass if you merge main into this branch.

**Comment by Shnatsel** [comment]
> Should `cargo-audit --color` support value `termcolor::ColorChoice::AlwaysAnsi`? Before I refactored `--color` flag, this value was not supported, so I left it that way. But should this remain?

I cannot think of any uses for it. We haven't seen any user requests for it either. Let's leave it out to keep the user interface simple, and only add it if someone asks for it.

> What about my question regarding `FixCommand`? I would like to propose some solution, so it can be discussed as `@tarcieri` suggested, but I would prefer to do it after current changes are accepted. Or should this be split off to another PR?

I think the bottom line is "none of the maintainers have a clue", so let's leave that to a follow-up PR.

> I would also like to do something with another issue I have created - https://github.com/rustsec/rustsec/issues/1161. As I understand, crate `platforms` is auto-generated, and so are values for target `Arch` and `OS`. But where exactly are they generated from? And how to point users to an exhaustive list of those values?

Yes, that is correct. I have explained where they come from in the linked issue. We should list the recognized values somewhere; either in the help text, or have a flag akin to `rustc --print=target-list`. Clap has long and short help texts; we could put the list into the long one I suppose? Anyway, that is also best to leave to another PR; I'd like to get this merged.

File: cargo-audit/audit.toml.example
Comment (Shnatsel) [inline cargo-audit/audit.toml.example:31]:
```suggestion
arch = ["x86_64"] # Ignore advisories for CPU architectures other than these
os = ["linux", "windows"] # Ignore advisories for operating systems other than these
```

File: cargo-audit/tests/audit.toml.legacy
Comment (Shnatsel) [inline cargo-audit/tests/audit.toml.legacy:4]: Excellent to have this tested :+1:

File: cargo-audit/tests/config.rs
Comment (Shnatsel) [inline cargo-audit/tests/config.rs:27]:
```suggestion
/// are specified as a string, and not a list. This is the legacy behavior.
```

**Review by Shnatsel — APPROVED** [review]
Looks good! 

The only issues I could spot are some typos in comments. Thanks a lot for this PR!

**Comment by Shnatsel** [comment]
I've rebased this myself in #1185 and merged that PR so that this doesn't stall out again.

Thanks a lot for your contribution!

I'd be happy to have more PRs from you, and hopefully I won't miss the notification next time.

**Pull request closed without merging**
